src/tests/end2end/IndexFormatTests.cpp

```diff
@@ -0,0 +1,268 @@
+                .SetAttribute(0, 0, nxt::VertexFormat::FloatR32G32B32A32, 0)
+                .GetResult();
+
+            nxt::ShaderModule vsModule = utils::CreateShaderModule(device, nxt::ShaderStage::Vertex, R"(
+                #version 450
+                layout(location = 0) in vec4 pos;
+                void main() {
+                    gl_Position = pos;
+                })"
+            );
+
+            nxt::ShaderModule fsModule = utils::CreateShaderModule(device, nxt::ShaderStage::Fragment, R"(
+                #version 450
+                layout(location = 0) out vec4 fragColor;
+                void main() {
+                    fragColor = vec4(0.0, 1.0, 0.0, 1.0);
+                })"
+            );
+
+            return device.CreateRenderPipelineBuilder()
+                .SetSubpass(renderpass, 0)
+                .SetPrimitiveTopology(nxt::PrimitiveTopology::TriangleStrip)
+                .SetStage(nxt::ShaderStage::Vertex, vsModule, "main")
+                .SetStage(nxt::ShaderStage::Fragment, fsModule, "main")
+                .SetIndexFormat(format)
+                .SetInputState(inputState)
+                .GetResult();
+        }
+};
+
+// Test that the Uint32 index format is correctly interpreted
+TEST_P(IndexFormatTest, Uint32) {
+    nxt::RenderPipeline pipeline = MakeTestPipeline(nxt::IndexFormat::Uint32);
+
+    nxt::Buffer vertexBuffer = utils::CreateFrozenBufferFromData<float>(device, nxt::BufferUsageBit::Vertex, {
+        -1.0f,  1.0f, 0.0f, 1.0f, // Note Vertices[0] = Vertices[1]
+        -1.0f,  1.0f, 0.0f, 1.0f,
+         1.0f,  1.0f, 0.0f, 1.0f,
+        -1.0f, -1.0f, 0.0f, 1.0f
+    });
+    // If this is interpreted as Uint16, then it would be 0, 1, 0, ... and would draw nothing.
+    nxt::Buffer indexBuffer = utils::CreateFrozenBufferFromData<uint32_t>(device, nxt::BufferUsageBit::Index, {
+        1, 2, 3
+    });
+
+    uint32_t zeroOffset = 0;
+    nxt::CommandBuffer commands = device.CreateCommandBufferBuilder()
+        .BeginRenderPass(renderpass, framebuffer)
+        .BeginRenderSubpass()
+            .SetRenderPipeline(pipeline)
+            .SetVertexBuffers(0, 1, &vertexBuffer, &zeroOffset)
+            .SetIndexBuffer(indexBuffer, 0)
+            .DrawElements(3, 1, 0, 0)
+        .EndRenderSubpass()
+        .EndRenderPass()
+        .GetResult();
+
+    queue.Submit(1, &commands);
+
+    EXPECT_PIXEL_RGBA8_EQ(RGBA8(0, 255, 0, 255), renderTarget, 100, 100);
+}
+
+// Test that the Uint16 index format is correctly interpreted
+TEST_P(IndexFormatTest, Uint16) {
+    nxt::RenderPipeline pipeline = MakeTestPipeline(nxt::IndexFormat::Uint16);
+
+    nxt::Buffer vertexBuffer = utils::CreateFrozenBufferFromData<float>(device, nxt::BufferUsageBit::Vertex, {
+        -1.0f,  1.0f, 0.0f, 1.0f,
+         1.0f,  1.0f, 0.0f, 1.0f,
+        -1.0f, -1.0f, 0.0f, 1.0f
+    });
+    // If this is interpreted as uint32, it will have index 1 and 2 be both 0 and render nothing
```

Comment:
The first index would be 0x00010002 or something, so wouldn't it potentially throw an error, failing to test the actual problem? (Although if the error got skipped because validation was done for Uint16, and then the driver received Uint32, then I imagine we'd either get garbage/crash or no rendering, so it's probably fine.)

Reply:
I'm hoping we get a crash there. I agree the tests aren't perfect but the other options would be to go full dEQP.